### PR TITLE
Guild Scheduled events

### LINKED ISSUE #938: Introduce discord guild scheduled event interface to hikari
### Summary

[Discord scheduled events](https://support.discord.com/hc/en-us/articles/4409494125719-Scheduled-Events) are a relatively new feature where server admins (and bots) can schedule events in a discord server. At the discord developer documentation, [there is a lot of api documentation for interacting with discord's RESTful API](https://discord.com/developers/docs/resources/guild-scheduled-event) with regards to scheduled events.

Relevant details from [discord documentation](https://discord.com/developers/docs/resources/guild-scheduled-event):

## Guild Scheduled Event Structure

Field | Type | Description
-- | -- | --
id | snowflake | the id of the scheduled event
guild_id | snowflake | the guild id which the scheduled event belongs to
channel_id | ?snowflake | the channel id in which the scheduled event will be hosted, or null if scheduled entity type is EXTERNAL
creator_id | ?snowflake | the id of the user that created the scheduled event
name | string | the name of the scheduled event (1-100 characters)
description? | string | the description of the scheduled event (1-1000 characters)
scheduled_start_time | ISO8601 timestamp | the time the scheduled event will start
scheduled_end_time | ?ISO8601 timestamp | the time the scheduled event will end, required if entity_type is EXTERNAL
privacy_level | privacy level | the privacy level of the scheduled event
status | event status | the status of the scheduled event
entity_type | scheduled entity type | the type of the scheduled event
entity_id | ?snowflake | the id of an entity associated with a guild scheduled event
entity_metadata | ?entity metadata | additional metadata for the guild scheduled event
creator? | user object | the user that created the scheduled event
user_count? | integer | the number of users subscribed to the scheduled event

### List Scheduled Events For Guild
```ml
GET /guilds/{guild.id}/scheduled-events/
```

Query String Params:

Field | Type | Description
-- | -- | --
with_user_count? | boolean | include number of users subscribed to each event

### Create Guild Scheduled Event
```ml
POST /guilds/{guild.id}/scheduled-events
```

JSON Params:

Field | Type | Description
-- | -- | --
channel_id? * | snowflake * | the channel id of the scheduled event.
entity_metadata? | entity metadata | the entity metadata of the scheduled event
name | string | the name of the scheduled event
privacy_level | privacy level | the privacy level of the scheduled event
scheduled_start_time | ISO8601 timestamp | the time to schedule the scheduled event
scheduled_end_time? | ISO8601 timestamp | the time when the scheduled event is scheduled to end
description? | string | the description of the scheduled event
entity_type | entity type | the entity type of the scheduled event

\* Optional for events with `'entity_type': EXTERNAL`

### Get Guild Scheduled Event

```ml
GET /guilds/{guild.id}/scheduled-events/{guild_scheduled_event.id}
```

Query String Params:

Field | Type | Description
-- | -- | --
with_user_count? | boolean | include number of users subscribed to this event

### Modify Guild Scheduled Event

```ml
PATCH /guilds/{guild.id}/scheduled-events/{guild_scheduled_event.id}
```

JSON Params:

Field | Type | Description
-- | -- | --
channel_id? * | ?snowflake | the channel id of the scheduled event, set to null if changing entity type to EXTERNAL
entity_metadata? | entity metadata | the entity metadata of the scheduled event
name? | string | the name of the scheduled event
privacy_level? | privacy level | the privacy level of the scheduled event
scheduled_start_time? | ISO8601 timestamp | the time to schedule the scheduled event
scheduled_end_time? * | ISO8601 timestamp | the time when the scheduled event is scheduled to end
description? | string | the description of the scheduled event
entity_type? * | event entity type | the entity type of the scheduled event
status? | event status | the status of the scheduled event

\* If updating `entity_type` to `EXTERNAL`:
* `channel_id` is required and must be set to null
* `entity_metadata` with a `location` field must be provided
* `scheduled_end_time` must be provided

### Delete Guild Scheduled Event

```ml
DELETE /guilds/{guild.id}/scheduled-events/{guild_scheduled_event.id}
```

### Get Guild Scheduled Event Users

```ml
GET /guilds/{guild.id}/scheduled-events/{guild_scheduled_event.id}/users
```

Query String Params:

Field | Type | Description | Default
-- | -- | -- | --
limit? | number | how many users to receive from the event | 100
with_member? | boolean | include guild member data if it exists | false
before? * | snowflake | consider only users before given user id | null
after? * | snowflake | consider only users after given user id | null

\* Provide a user id to before and after for pagination. Users will always be returned in ascending order by `user_id`. If both `before` and `after` are provided, only `before` is respected. Fetching users in-between `before` and `after` is not supported.

### Why is this needed?

> With Scheduled events, you will now be able to schedule "events" in your server ahead of time so that your community can learn about and be notified of when they're occurring and be notified when you're live. [-discord](https://support.discord.com/hc/en-us/articles/4409494125719-Scheduled-Events)

I think that this is needed in order to fill in some holes in the current API that `hikari` offers. 

Scheduled events are useful for organizing community activities, although there are other discord api libraries that offer similar features ([see disnake documentation for examples.](https://docs.disnake.dev/en/latest/api.html?highlight=guild#disnake.Guild.create_scheduled_event))

Specifically in my use case, I am making a bot which allows users to create server-wide events if someone pings a role, so we can use discord's native interface to RSVP for the event.

### Ideal implementation

To be decided.

Disnake's implementation of this is [here](https://github.com/DisnakeDev/disnake/blob/master/disnake/guild_scheduled_event.py) (they use an MIT license).

I think that at the very least, the `Discord Scheduled Event Structure` should be implemented with all of its fields, and all of the HTTP requests I listed previously are also implemented in some way.

### Checklist

- [X] I have searched the issue tracker and have made sure it's not a duplicate. If it is a follow up of another issue, I have specified it.

### PR DESCRIPTION
**Pull request by FasterSpeeding**
### Summary
Guild Scheduled events

### Checklist

- [x] I have run `nox` and all the pipelines have passed.
- [x] I have made unittests according to the code I have added/modified/deleted.

### Related issues
Closes #938

### CODE REVIEW

File: hikari/impl/entity_factory.py [hunk 1]
```diff
@@ -2683,6 +2690,117 @@ def deserialize_member_presence(  # noqa: CFQ001 - Max function length
...
+
+        return scheduled_events_models.ScheduledExternalEvent(
+            app=self._app,
+            id=snowflakes.Snowflake(payload["id"]),
+            guild_id=snowflakes.Snowflake(payload["guild_id"]),
+            name=payload["name"],
+            description=payload.get("description"),
+            start_time=time.iso8601_datetime_string_to_datetime(payload["scheduled_start_time"]),
+            end_time=time.iso8601_datetime_string_to_datetime(payload["scheduled_end_time"]),
+            privacy_level=scheduled_events_models.EventPiracyLevel(payload["privacy_level"]),
```
Comment (HyperGH) [inline hikari/impl/entity_factory.py:2712]:
```suggestion
            privacy_level=scheduled_events_models.EventPrivacyLevel(payload["privacy_level"]),
```

File: hikari/impl/entity_factory.py [hunk 2]
```diff
@@ -2683,6 +2690,117 @@ def deserialize_member_presence(  # noqa: CFQ001 - Max function length
...
+
+        return scheduled_events_models.ScheduledStageEvent(
+            app=self._app,
+            id=snowflakes.Snowflake(payload["id"]),
+            guild_id=snowflakes.Snowflake(payload["guild_id"]),
+            name=payload["name"],
+            description=payload.get("description"),
+            start_time=time.iso8601_datetime_string_to_datetime(payload["scheduled_start_time"]),
+            end_time=end_time,
+            privacy_level=scheduled_events_models.EventPiracyLevel(payload["privacy_level"]),
```
Comment (HyperGH) [inline hikari/impl/entity_factory.py:2740]:
```suggestion
            privacy_level=scheduled_events_models.EventPrivacyLevel(payload["privacy_level"]),
```

File: hikari/scheduled_events.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Application and entities that are used to describe guild scheduled events on Discord."""
+from __future__ import annotations
+
+__all__: typing.Sequence[str] = [
+    "EventPiracyLevel",
```
Comment (HyperGH) [inline hikari/scheduled_events.py:27]:
```suggestion
    "EventPrivacyLevel",
```

File: hikari/scheduled_events.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+if typing.TYPE_CHECKING:
+    import datetime
+
+    from hikari import files
+    from hikari import guilds
+    from hikari import traits
+    from hikari import users
+
+
+class EventPiracyLevel(int, enums.Enum):
```
Comment (HyperGH) [inline hikari/scheduled_events.py:56]:
```suggestion
class EventPrivacyLevel(int, enums.Enum):
```

File: hikari/scheduled_events.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    """Enum of the possible scheduled event piracy levels."""
```
Comment (HyperGH) [inline hikari/scheduled_events.py:57]:
```suggestion
    """Enum of the possible scheduled event privacy levels."""
```

File: hikari/scheduled_events.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    description: typing.Optional[str] = attr.field(hash=False, repr=False)
+    """Description of the scheduled event."""
+
+    start_time: datetime.datetime = attr.field(hash=False, repr=False)
+    """When the event is scheduled to start."""
+
+    end_time: typing.Optional[datetime.datetime] = attr.field(hash=False, repr=False)
+    """When the event is scheduled to end, if set."""
+
+    privacy_level: EventPiracyLevel = attr.field(hash=False, repr=False)
```
Comment (HyperGH) [inline hikari/scheduled_events.py:126]:
```suggestion
    privacy_level: EventPrivacyLevel = attr.field(hash=False, repr=False)
```

File: hikari/events/scheduled_events.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+"""Events fired for guild scheduled event related changes."""
+from __future__ import annotations
+
+__all__: typing.Sequence[str] = [
+    "ScheduledEventEvent",
+    "ScheduledEventCreateEvent",
+    "ScheduledEventDeleteEvent",
+    "ScheduledEventUpdateEvent",
+    "ScheduledEventUserAddEvent",
+    "ScheduledEventUserRemoveEvent",
```
Comment (davfsa) [inline hikari/events/scheduled_events.py:32]:
```suggestion
    "ScheduledEventUserSubscribeEvent",
    "ScheduledEventUserUnsubscribeEvent",
```
Make them a bit longer, but better represent the event

File: hikari/scheduled_events.py (hunk 5 above)
Comment (davfsa) [inline hikari/scheduled_events.py:57]:
```suggestion
    """Enum of the possible scheduled event privacy levels."""
```
Might be missed, so just leaving it here to be sure

File: changes/idk.feature.md [hunk 8]
```diff
@@ -0,0 +1 @@
+Scheduled event support.
```
Comment (davfsa) [inline changes/idk.feature.md:1]: File needs renaming

**Review by davfsa — DISMISSED** [review]
Merge conflicts, but LGTM!

**Review by davfsa — APPROVED** [review]

**Pull request merged**